numpy/lib/npyio.py

```diff
@@ -1033,23 +1033,25 @@ def read_data(lineno_words_iter, chunk_size):
         byte_converters = True
 
     if usecols is not None:
-        # Allow usecols to be a single int or a sequence of ints
+        # Copy usecols, allowing it to be a single int or a sequence of ints.
         try:
-            usecols_as_list = list(usecols)
+            usecols = list(usecols)
         except TypeError:
-            usecols_as_list = [usecols]
-        for col_idx in usecols_as_list:
+            usecols = [usecols]
+        for i, col_idx in enumerate(usecols):
             try:
-                opindex(col_idx)
+                usecols[i] = opindex(col_idx)  # Cast to builtin int now.
             except TypeError as e:
                 e.args = (
                     "usecols must be an int or a sequence of ints but "
                     "it contains at least one element of type %s" %
                     type(col_idx),
                     )
                 raise
-        # Fall back to existing code
-        usecols = usecols_as_list
+        usecols_getter = (
+            itemgetter(*usecols) if len(usecols) > 1 else
+            # Get an iterable back, even if using a single column.
+            lambda obj, _col=usecols[0]: [obj[_col]])
 
```

Comment:
It would be easier to parse this code if this was moved up into the `read_data` function that is the only place it is used. It could then be handled with the usual if statements.

I must say this function is a mess, I salute you for wading through it.

Reply:
Can you review https://github.com/numpy/numpy/pull/19693 first instead then? I guess that would be the more logical order to do things.

Reply:
Rebased (thanks for the quick merge of #19693), and slightly rewrote the logic, but I think it's still logical to define usecols_getter at the place where usecols is getting parsed, and have a single if check (now just a None-comparison) in the main loop?  The point being of course to minimize the amount of conditional logic in the main loop...

FTR, running all loadtxt benchmarks after this rebase gives
```
       before           after         ratio
     [1fdfdfa9]       [a59c317b]
     <main>           <loadtxtusecols>
-     5.53±0.05ms      5.02±0.07ms     0.91  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv(2)
-     9.46±0.05ms      8.46±0.07ms     0.89  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv([1, 3, 5, 7])
-     7.78±0.05ms      6.86±0.03ms     0.88  bench_io.LoadtxtUseColsCSV.time_loadtxt_usecols_csv([1, 3])
```

Reply:
My would use ordinary `if` statements to define `usecols_getter` rather than a lengthy right hand side. It would be easier to read and parse.

Reply:
Ah, sure.